src/dot-protocol/merlin_dot_protocol.ml

```diff
@@ -127,39 +127,74 @@ module Sexp = struct
     List (List.map ~f directives)
 end
 
-module Csexp = Csexp.Make (Sexp)
+type read_error =
+  | Unexpected_output of string
+  | Csexp_parse_error of string
 
-module Commands = struct
-  type t = File of string | Halt | Unknown
+module Make (IO : sig
+    type 'a t
+
+    module O : sig
+      val ( let+ ) : 'a t -> ('a -> 'b) -> 'b t
+    end
+  end) (Chan : sig
+          type in_chan
+          type out_chan
+
+          val read : in_chan -> (Csexp.t, string) result IO.t
+
+          val write : out_chan -> Csexp.t -> unit IO.t
+        end) = struct
+  module Commands = struct
+    type t = File of string | Halt | Unknown
+
+    let read_input chan =
+      let open Sexp in
+      let open IO.O in
+      let+ input = Chan.read chan in 
+      match input with
+      | Ok (List [Atom "File"; Atom path]) -> File path
+      | Ok (Atom "Halt") -> Halt
+      | Ok _ -> Unknown
+      | Error _ -> Halt
+
+    let send_file chan path =
+      Chan.write chan Sexp.(List [Atom "File"; Atom path])
+
+    let halt chan =
+      Chan.write chan (Sexp.Atom "Halt")
+  end
 
-  let read_input in_channel =
-    let open Sexp in
-    match Csexp.input in_channel with
-    | Ok (List [Atom "File"; Atom path]) -> File path
-    | Ok (Atom "Halt") -> Halt
-    | Ok _ -> Unknown
-    | Error _msg -> Halt
+  let read chan =
+    let open IO.O in
+    let+ res = Chan.read chan in
+    match res with
+    | Ok (Sexp.List directives) ->
+      Ok (List.map directives ~f:Sexp.to_directive)
+    | Ok sexp ->
+      let msg = Printf.sprintf
+          "A list of directives was expected, instead got: \"%s\""
+          (Sexp.to_string sexp)
+      in
+      Error (Unexpected_output msg)
+    | Error msg -> Error (Csexp_parse_error msg)
 
-  let send_file ~out_channel path =
-    Sexp.(List [Atom "File"; Atom path])
-    |> Csexp.to_channel out_channel
+  let write out_chan (directives : directive list) =
+    directives |> Sexp.from_directives |> Chan.write out_chan
 end
 
-type read_error =
-  | Unexpected_output of string
-  | Csexp_parse_error of string
+module StdIO = struct
+  type 'a t = 'a
+  module O = struct
+    let ( let+ ) x f = f x
+  end
+end
 
-let read ~in_channel =
-  match Csexp.input in_channel with
-  | Ok (Sexp.List directives) ->
-      Ok (List.map directives ~f:Sexp.to_directive)
-  | Ok sexp ->
-    let msg = Printf.sprintf
-      "A list of directives was expected, instead got: \"%s\""
-      (Sexp.to_string sexp)
-    in
-    Error (Unexpected_output msg)
-  | Error msg -> Error (Csexp_parse_error msg)
+module StdChan = struct
+  type in_chan = in_channel
+  type out_chan = out_channel
+  let read = Csexp.input
+  let write = Csexp.to_channel
+end
 
-let write ~out_channel (directives : directive list) =
-  directives |> Sexp.from_directives |> Csexp.to_channel out_channel
+module Std = Make(StdIO)(StdChan)
```

Comment:
I think the most important feature of this instance is that it is blocking. A name like `Blocking` or `Blocking_std` might be better.

Also  let's inline the arguments, they are small enough for that.

Reply:
Done, used `Blocking`